packages/api-request/src/middlewares/nonce.js

```diff
@@ -0,0 +1,51 @@
+/**
+ * External dependencies
+ */
+import jQuery from 'jquery';
+
+const createNonceMiddleware = ( nonce ) => ( options, next ) => {
+	let usedNonce = nonce;
+	/**
+	 * This is not ideal but it's fine for now.
```

Comment:
It didn't feel right to me to tie `apiRequest`/fetch to heartbeat specific events. I think it's better in the last versions using hooks. I don't have an ideal solution without complexity though.

Reply:
Looking at it some more, I think I could see a couple things:

- There's an `X-WP-Nonce` on the response of a fetch request; should we not assume that as a fresh one and use it?
- Maybe this (the heartbeat tick handling) should be pulled out and added as an inline script in the WordPress-specific usage of `api-request`, with some way to _inform_ the middleware of a new nonce.

Reply:
Both these suggestion seem good to me, I just worry about the complexity especially about this:

> with some way to inform the middleware of a new nonce.

If we manage to keep it simple in terms of API and implementation, all good for me.

Reply:
One idea: optionally allow, instead of an initial nonce, a function which receives as an argument the "inform" callback:

```js
createNonceMiddleware( ( setNonce ) => {
	setNonce( initialNonce );

	addAction(
		'heartbeat.tick',
		'core/api-fetch/create-nonce-middleware',
		( response ) => response[ 'rest_nonce' ] && setNonce( response[ 'rest_nonce' ] ),
	);
} );
```

Lazy man's observable.

Reply:
> There's an X-WP-Nonce on the response of a fetch request; should we not assume that as a fresh one and use it?

With that in mind, should there be an optional callback that is passed in that can be used to update the nonce when its observed in the response of a fetch?

Reply:
Or, just have the nonce as a property of the created middleware.

```js
const nonceMiddleware = createNonceMiddleware( initialNonce );
wp.apiFetch.use( nonceMiddleware );
wp.hooks.addAction(
	'heartbeat.tick', 
	'core/api-fetch/create-nonce-middleware',
	( response ) => response[ 'rest_nonce' ] && nonceMiddleware.nonce = response[ 'rest_nonce' ]
);
```

Reply:
>With that in mind, should there be an optional callback that is passed in that can be used to update the nonce when its observed in the response of a fetch?

What's the use-case?

Reply:
I was thinking there might be some global reference to the _current_ nonce but I'm guessing there's not (and doesn't need to be obviously).

Reply:
I think of the options given, I prefer the callback approach.  Exposing nonce as a property on nonceMiddleware seems potentially fragile because third party code could clobber it right?

Reply:
They'd need a reference to the created middleware (not likely to be exposed). Could even be argued to be a good thing to allow someone to update it themselves if they want, though. I don't see it likely to be an issue.